src/main/scala/com/microsoft/hyperspace/index/plananalysis/FilterReason.scala

```diff
@@ -0,0 +1,191 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.plananalysis
+
+trait FilterReason {
+  val args: Seq[(String, String)]
+  def codeStr: String
+  final def argStr: String = {
+    // key1=[value1], key2=[value2]
+    args.map(kv => s"${kv._1}=[${kv._2}]").mkString(",")
+  }
+  def verboseStr: String
+}
+
+trait FilterReasonNoArg extends FilterReason {
+  final override val args: Seq[(String, String)] = Seq.empty
+}
+
+object FilterReasonCode extends Enumeration {
+  type FilterReasonCode = Value
+
+  // Common
+  val COL_SCHEMA_MISMATCH = Value
+  val SOURCE_DATA_CHANGED = Value
+  val NO_DELETE_SUPPORT = Value
+  val NO_COMMON_FILES = Value
+  val TOO_MUCH_APPENDED = Value
+  val TOO_MUCH_DELETED = Value
+  val ANOTHER_INDEX_APPLIED = Value
+
+  // CoveringIndex - FilterIndexRule
+  val NO_FIRST_INDEXED_COL_COND = Value
+  val MISSING_REQUIRED_COL = Value
+
+  // CoveringIndex - JoinIndexRule
+  val NOT_ELIGIBLE_JOIN = Value
+  val NO_AVAIL_JOIN_INDEX_PAIR = Value
+  val NO_COMPATIBLE_JOIN_INDEX_PAIR = Value
+  val NOT_ALL_JOIN_COL_INDEXED = Value
+  val MISSING_INDEXED_COL = Value
+}
+
+object FilterReasons {
+  import com.microsoft.hyperspace.index.plananalysis.FilterReasonCode._
+  def apply(code: FilterReasonCode, args: (String, String)*): FilterReason = {
+    code match {
+      case COL_SCHEMA_MISMATCH =>
+        ColSchemaMismatch(args)
+      case SOURCE_DATA_CHANGED =>
+        SourceDataChanged()
+      case NO_DELETE_SUPPORT =>
+        NoDeleteSupport()
+      case NO_COMMON_FILES =>
+        NoCommonFiles()
+      case TOO_MUCH_APPENDED =>
+        TooMuchAppended(args)
+      case TOO_MUCH_DELETED =>
+        TooMuchDeleted(args)
+      case MISSING_REQUIRED_COL =>
+        MissingRequiredCol(args)
+      case NO_FIRST_INDEXED_COL_COND =>
+        NoFirstIndexedColCond(args)
+      case NOT_ELIGIBLE_JOIN =>
+        NotEligibleJoin(args)
+      case NO_AVAIL_JOIN_INDEX_PAIR =>
+        NoAvailJoinIndexPair(args)
+      case MISSING_INDEXED_COL =>
+        MissingIndexedCol(args)
+      case NOT_ALL_JOIN_COL_INDEXED =>
+        NotAllJoinColIndexed(args)
+      case NO_COMPATIBLE_JOIN_INDEX_PAIR =>
+        NoCompatibleJoinIndexPair()
+      case ANOTHER_INDEX_APPLIED =>
+        AnotherIndexApplied(args)
+    }
+  }
+
+  case class ColSchemaMismatch(override val args: Seq[(String, String)])
+      extends FilterReason {
+    override final val codeStr: String = "COL_SCHEMA_MISMATCH"
+    override def verboseStr: String = {
+      "Column Schema does not match. Source data columns: [" + args(0)._2 +
+        "], Index columns: [" + args(1)._2
+    }
+  }
```

Comment:
How about creating ColSchemaMismatch and other reasons directly instead of doing it indirectly with FilterReasons.apply? Using `Seq[(String, String)]` and FilterReasons.apply seems to make things more complex and hard to use than they have to be.

```scala
// reason definition
case class ColSchemaMismatch(sourceColumns: String, indexColumns: String) extends FilterReason {
  override final val codeStr: String = "COL_SCHEMA_MISMATCH"
  override def args: Seq[(String, String)] = Seq("sourceColumns" -> sourceColumns, "indexColumns" -> indexColumns)
  override def verboseStr: String =
    s"Column Schema does not match. Source data columns: [$sourceColumns], Index columns: [$indexColumns]"
}

// reason creation example
ColSchemaMismatch(relationColumnNames.mkString(", "), index.derivedDataset.referencedColumns.mkString(", "))

// reason creation example 2
ColSchemaMismatch(
  sourceColumns = relationColumnNames.mkString(", "),
  indexColumns = index.derivedDataset.referencedColumns.mkString(", "))
```

Reply:
This way looks clearer 👍